### DISCUSSION TITLE
usbh_edpt_xfer_with_callback

### DISCUSSION THREAD

**Discussion by abakosh** [hathach/tinyusb#2118] [General]
Hello @hathach , 

I am using the host module for MSC on Renesas RA, with no rtos

I want to know what is the common reason for the following assertion to fail.

**usbh_edpt_xfer_with_callback 797: ASSERT FAILED**

I am still not sure why I end up with this error message, the problem is when I get into this, It freezes my program completely, I think there is something wrong with the hcd layer, but I am not quite sure.

Here is a snippet of what I am getting at level one debug (I enabled Queue EP manually)

  Queue EP 81 with 13 bytes ... 
  Queue EP 02 with 31 bytes ... 
  Queue EP 81 with 512 bytes ... 
  Queue EP 81 with 13 bytes ... 
  Queue EP 02 with 31 bytes ... 
  Queue EP 81 with 512 bytes ... 
  Queue EP 81 with 13 bytes ... 
  Queue EP 02 with 31 bytes ... 
  Queue EP 81 with 512 bytes ... 
  Queue EP 81 with 13 bytes ... 
  Queue EP 02 with 31 bytes ... 
  Queue EP 81 with 512 bytes ... 
  Queue EP 81 with 13 bytes ... 
  Queue EP 02 with 31 bytes ... 
  Queue EP 81 with 512 bytes ... 
  Queue EP 81 with 13 bytes ... 
  Queue EP 02 with 31 bytes ... 
  Queue EP 81 with 512 bytes ... 
  Queue EP 81 with 13 bytes ... 
usbh_edpt_xfer_with_callback 797: ASSERT FAILED
msch_xfer_cb 352: ASSERT FAILED

It freezes here

**Comment by abakosh** [comment]
@hathach I found a clear issue in the hcd  of Renesas, its in nrdy handling, the pid flag can be 2 or 3 if it is stalling, however the handler is not considering 3 at all , and go to default in the switch statement . 

  switch (*ctr & RUSB2_PIPE_CTR_PID_Msk) { 
     default: return; 
     case RUSB2_PIPE_CTR_PID_STALL: result = XFER_RESULT_STALLED; break; 
     case RUSB2_PIPE_CTR_PID_NAK:   result = XFER_RESULT_FAILED;  break; 
   }

**Comment by mcuenthusiast** [comment]
I was facing the NRDY issue while using the MSC code with renesas RA6 board compiled on IAR.

After making below changes my code worked you can also try if you still have a issue.

In exisiting code under below function.
pipe_write_packet() -- MBW bit of CFIFOSEL register is set to 16-bit access by default, even if it is only writing 1 byte of data.

So I changed to to something like below.

<img width="448" alt="image" src="https://github.com/hathach/tinyusb/assets/132841642/ed2e3635-5377-417a-b031-ef297f3973b3">

**Reply by abakosh** [reply]
Thank you for your reply, I don't think this will fix it. please refer to the UM
![image](https://github.com/hathach/tinyusb/assets/36272850/1096833b-ce26-45f3-aeaf-7856d09501af)
You might fixed something somewhere else. are you sure these are the only changes you did?
